feat: new transaction feature

danieljbruce opened the pull request:
**Summary**

This PR ensures that transaction read calls will use [newTransaction options](https://github.com/googleapis/nodejs-datastore/blob/1585d4a4e1b4b16d198307a3e97ffcf156d000b1/protos/protos.d.ts#L7588) to begin transactions instead of making a separate `beginTransaction` grpc call.

**Source code changes**

A `#blockWithMutex` function is added and it is used for read calls instead of a `#withBeginTransaction` function. This makes it so that all read calls on Transaction will block on the mutex and then call the same method in their super class. Before this change, the read calls on Transaction would block on the mutex, make a `beginTransaction` grpc call and then call the same method in their super class. So this change to using `#blockWithMutex` omits the extra `beginTransaction` call and instead expects the function it is calling to supply new transaction options.

A function `getTransactionRequest` is added which pulls out the code used for building transaction options for a `beginTransaction` request. That is because this code is also used to build the new transaction options.

TransactionState is moved up to the super class `DatastoreRequest` and a new state is added called `NOT_TRANSACTION` for `DataRequests` that are not transactions.

`parseTransactionResponse` is moved up to the super class `DatastoreRequest` and is now used by `runQuery`, `runAggregationQuery` and `get` to save the transaction id.

`newTransaction` options are now built in `getRequestOptions` which is used by `runQuery`, `runAggregationQuery` and `get` to save the transaction id.

`state` is moved up to the `DatastoreRequest` super class because it now has to be used to decide whether or not to build newTransaction options.

If a transaction is expired then an error will be thrown when the transaction is used.

ReadOptions are the options shared by Datastore read requests.
[new_transaction](https://github.com/googleapis/go-genproto/blob/478b75c54725e54e5f92966212c6572c091596aa/googleapis/datastore/v1/datastore.pb.go#L1794-L1803) is a [consistency type](https://github.com/googleapis/go-genproto/blob/478b75c54725e54e5f92966212c6572c091596aa/googleapis/datastore/v1/datastore.pb.go#L1708) in ReadOptions. When new_transaction is used in a read request, a new transaction is begun for the request.

**System tests added**

A system test is added to ensure using new transaction has better latency. System tests are added for runQuery, runAggregationQuery and get for readOnly transactions to ensure readOnly transactions work correctly whether `transaction.run` is called or not.

**Unit tests added**

Unit tests are added for various orders of transaction operations that ensure correct behaviour for when `transaction.run` is used and when `transaction.run` is omitted. `lookup, lookup, put, commit` is one order that is tested. The others are `runQuery, lookup, put, commit`, `runAggregationQuery, lookup, put, commit`, `put, put, lookup, commit` and `put, commit`. A unit test is also corrected.

**Drawbacks**

If the user uses `transaction.runQueryStream` or they use `transaction.createReadStream` then the call will not block on a mutex. These methods are not async methods so they cannot wait for a `#blockWithMutex` call to complete. If users are currently using `runQueryStream` or `createReadStream` concurrently then these multiple read calls may be beginning multiple transactions at the same time using one transaction object which is not ideal.

danieljbruce committed:
Move the state up, make it protected

Transaction state should move up to the super class

danieljbruce committed:
Build new transaction options method

Add a method for building transaction options. Also add a method for detecting if a request is a transaction. Build new transaction options

danieljbruce committed:
Correct the fn that builds transactions

Corrected getTransactionRequest. Build the request options properly.

danieljbruce committed:
Set transaction to readOnly in system test.

When this system test isn’t set to be a readOnly transaction then the error occurs which says too much contention.

danieljbruce committed:
Change #parseRunSuccess

Change parseRunSuccess so that it can be used more universally.

danieljbruce committed:
Move parseRunSuccess up to protected

parseRunSuccess should move up to the super class because it now needs to be used there.

danieljbruce committed:
Saves the transaction id

This change saves the transaction id returned from the server for read calls

danieljbruce committed:
Add tests for testing read only transactions

Read only tests are needed

danieljbruce committed:
Add tests that measure read time

Make sure that code using the new transaction option has better performance than code that doesn’t have it.

danieljbruce committed:
ran linter

danieljbruce committed:
Add a test for testing requests

Use the MockedTransactionWrapper to test requests being passed into the Gapic layer.

danieljbruce committed:
Final changes to make test work

Mock out begin transaction. Test for newTransaction consistency type. Mock out commit.

danieljbruce committed:
Add the transaction.run test

A transaction.run test is needed  for lookup, lookup, put, commit.

danieljbruce committed:
run linter

danieljbruce committed:
runQuery, lookup, put, commit

Add a test for this sequence of operations and ensure it works properly.

danieljbruce committed:
runAggregationQuery, lookup, put, commit

Four operations that get all the results for running an aggregation query. Adding another test for these four operations.

danieljbruce committed:
put, put, lookup, commit

Last test suite regarding new transaction unit tests.

danieljbruce committed:
Add tests for the commits

Add a bunch of tests for the commit case. Check the commit gapic input.

danieljbruce committed:
Add testing to ensure begin tx is called

Begin transaction should be called at least once. Add code here to increment the counter.

danieljbruce committed:
Document #blockWithMutex

Remove commented code too.

danieljbruce committed:
Document transaction state

danieljbruce committed:
feat: new transaction feature branch

danieljbruce committed:
Add a check for expired

Check for expired on most functions and write tests for the expired check.

danieljbruce committed:
Add commit and rollback blocks

Check for expired state in the commit and rollback blocks.

danieljbruce committed:
run the linter

danieljbruce commented on src/request.ts:
```diff
@@ -1116,6 +1188,36 @@ class DatastoreRequest {
+  return request instanceof Transaction;
+}
+
+export function getTransactionRequest(
```
This function is used by `DatastoreRequest` in `request.ts` and `Transaction` in `transaction.ts`. `Transaction` inherits from `DatastoreRequest` so even though this could be moved to a protected method in `DatastoreRequest`, we want to avoid doing so. This is because adding the `protected` reserved word in Typescript doesn't entirely hide the method from the user in the class like `protected` should.

danieljbruce commented on src/request.ts:
```diff
@@ -89,6 +109,7 @@ class DatastoreRequest {
     | Array<(err: Error | null, resp: Entity | null) => void>
     | Entity;
   datastore!: Datastore;
+  protected state: TransactionState = TransactionState.NOT_TRANSACTION;
```
I don't like using `protected` here because typescript doesn't technically hide `protected` properties completely, but it seems that there is no simple way around it because both `Transaction` and `DatastoreRequest` need access to the `state` variable. There is more information about some complex ways to avoid this in the design doc in Alternative 3, but they require major architecture changes.

danieljbruce committed:
Don’t allow readtime to be specified in a txn

Matches the python client.

danieljbruce committed:
throw error for both read time and consistency

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-datastore into new-transaction-feature-branch

# Conflicts:
#	src/request.ts

danieljbruce committed:
Add test for specifying readtime

danieljbruce committed:
Remove the console logs

danieljbruce committed:
Refactor the test

danieljbruce committed:
Improve test to make sure Gapic layer isn’t called

Also, fix bug with read consistency.

danieljbruce committed:
Run linter
Reorganize tests
New test for readtime and consistency

danieljbruce committed:
Make change to allow the runAggregationQuery go up

danieljbruce committed:
Should error when get is used

danieljbruce committed:
Move tests over and refactor initialized datastore

danieljbruce committed:
Remove only and remove imports

danieljbruce committed:
Introduce parameterized testing for errors

danieljbruce committed:
Use parameters in parameterized tests

danieljbruce committed:
Migrate error tests to parameterized testing

danieljbruce committed:
Change description

danieljbruce committed:
Prepare second describe block for tests

danieljbruce committed:
Revert "Prepare second describe block for tests"

This reverts commit 974117b05b4e387a4e34668651967641bbde501c.

danieljbruce committed:
Always return after the error is sent back

This ensures that an extra call does not get made to the server.

danieljbruce commented on src/request.ts:
```diff
@@ -990,16 +1061,34 @@ class DatastoreRequest {
       const code = CONSISTENCY_PROTO_CODE[options.consistency.toLowerCase()];
-      sharedQueryOpts.readOptions = {
-        readConsistency: code,
-      };
```
This was a bug because it would overwrite `newTransaction` if `readConsistency` was provided.

danieljbruce commented on src/request.ts:
```diff
@@ -1122,18 +1211,27 @@ class DatastoreRequest {
-          'Read consistency cannot be specified in a transaction.'
-        );
-      }
-
```
This is just moved up to also be included for new transaction calls. Also, the error was not propogating up and needs to be wrapped in a callback.

danieljbruce commented on src/transaction.ts:
```diff
@@ -53,11 +54,6 @@ interface BeginAsyncResponse {
-  NOT_STARTED,
-  IN_PROGRESS, // IN_PROGRESS currently tracks the expired state as well
-}
-
```
Moved to the super class because it is now needed there.

danieljbruce commented on src/transaction.ts:
```diff
@@ -696,24 +684,10 @@ class Transaction extends DatastoreRequest {
     return new Promise((resolve: (value: BeginAsyncResponse) => void) => {
+      const reqOpts = {
+        transactionOptions: getTransactionRequest(this, options),
+      };
```
`getTransactionRequest` does what the removed code does here.

danieljbruce committed:
Add headers

danieljbruce committed:
Tests and implementation for expired on rollback

danieljbruce committed:
fix test

danieljbruce committed:
Wrap rollbacks with a withBeginTransaction.

danieljbruce committed:
Update the test so that it begins the tx before

danieljbruce committed:
Throw error if transaction not started on rollback

danieljbruce committed:
Remove only

danieljbruce committed:
Add a comment to the test regarding new txn.run()

danieljbruce committed:
Ensure that the errors get bubbled up

danieljbruce committed:
Run. linter

danieljbruce committed:
Merge branch 'main' into new-transaction-feature-branch

danieljbruce committed:
read time and consistency error

danieljbruce committed:
Merge branch 'new-transaction-feature-branch' of https://github.com/danieljbruce/nodejs-datastore into new-transaction-feature-branch

danieljbruce committed:
Remove unnecessary change

danieljbruce committed:
Eliminate the call to withBeginTransaction

danieljbruce commented on test/gapic-mocks/runQuery.ts:
```diff
@@ -68,35 +53,13 @@ describe('Run Query', () => {
     }
   }
 
-  it('should pass read time into runQuery for transactions', async () => {
```
Note that this test is no longer valid because transaction and read time are no longer compatible together as is the case in Python.

cindy-peng commented on src/transaction.ts:
```diff
@@ -434,16 +430,32 @@ class Transaction extends DatastoreRequest {
+      callback(new Error(transactionExpiredError));
+      return;
+    }
+    if (this.state === TransactionState.NOT_STARTED) {
```
Just wondering what are the allowed states here?  NOT_TRANSACTION and IN_PROGRESS?

danieljbruce replied:
NOT_TRANSACTION, NOT_STARTED, IN_PROGRESS, EXPIRED.

cindy-peng commented on src/request.ts:
```diff
@@ -846,6 +929,10 @@ class DatastoreRequest {
+    if (this.state === TransactionState.EXPIRED) {
+      callback(new Error(transactionExpiredError));
+      return;
+    }
```
Do we need to check `readTimeAndConsistencyError` here?

danieljbruce replied:
No. That check is done further downstream in the call to `runQueryStream`. But I did spot some other cases where errors weren't bubbling up and fixed those bugs.

danieljbruce committed:
Throw error reported in documentation

danieljbruce commented on src/request.ts:
```diff
@@ -335,6 +366,15 @@ class DatastoreRequest {
+
   /**
    * Retrieve the entities as a readable object stream.
    *
```
Sure. Done.

danieljbruce committed:
Work on streaming errors

Fix the concurrency tests and move the error reporting outside of the stream, add two tests for the streams to make sure errors get reported.

danieljbruce committed:
Move the error throwing up the stack

This makes it so that it is reported in the streams.

danieljbruce committed:
Fix linting errors

daniel-sanche reviewed (approved):
LGTM

danieljbruce committed:
Generate unit tests for getTransactionRequest

I am going to rewrite the getTransactionRequest function so it is a good idea not to change how it works.

danieljbruce committed:
Generate unit tests

Fix the compiler error
Change code so that when readOnly is specified then readwrite options are ignored.
Fix the unit test to adopt this change.

danieljbruce committed:
Merge branch 'main' into new-transaction-feature-branch

danieljbruce committed:
Make function one ternary operation

Add comments, simplify logic in the function.

danieljbruce committed:
Merge branch 'new-transaction-feature-branch' of https://github.com/danieljbruce/nodejs-datastore into new-transaction-feature-branch

danieljbruce committed:
Add two comments

danieljbruce commented on src/request.ts:
```diff
@@ -1229,6 +1328,54 @@ class DatastoreRequest {
+    if (id) {
+      reqOpts.readWrite = {previousTransaction: id};
+    }
+  }
```
I generated unit tests, corrected them and then reduced this function down to one ternary operation. Previously this was just copy pasted code from before. Search `describe('getTransactionRequest', () => {` to see how this function should behave.

danieljbruce commented on src/request.ts:
```diff
@@ -1229,6 +1328,54 @@ class DatastoreRequest {
+export function getTransactionRequest(
+  transaction: Transaction,
+  options: RunOptions
+): TransactionRequestOptions {
```
Done

danieljbruce committed:
Update the comment

danieljbruce committed:
Run the linter
